### ISSUE TITLE
Race condition in PENestedNamespaceSymbol.EnsureAllMembersLoaded

### ISSUE THREAD

**Issue by AntonLapounov** [dotnet/roslyn#34297]
https://github.com/dotnet/roslyn/blob/fab7134296816fc80019c60b0f5bef7400cf23ea/src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENestedNamespaceSymbol.cs#L122-L132

Reading `_typesByNS` at line 124 is not synchronized with writing it at line 130, which creates a race condition.  Namely, reading `_typesByNS` may be reordered with checking the `lazyTypes == null || lazyNamespaces == null` condition.  If another thread sets `_typesByNS` to `null` in that window, this thread then passes `null` to `LoadAllMembers`, which crashes the process due to an unhandled `NullReferenceException`.

**Comment by AntonLapounov** [comment]
@AlekseyTs, please help to route this.

**Comment by AlekseyTs** [comment]
@jaredpar For triage

**Comment by jaredpar** [comment]
@AntonLapounov thanks for reporting. Does seem like a bug here. What context are you hitting this in?

**Comment by jaredpar** [comment]
@gafter who do we have this indirect check for `null` here vs. just checking `typesNs` directly?

**Comment by AntonLapounov** [comment]
@jaredpar We have collected several Roslyn crash dumps in .NET Native automated runs, where ILC-compiled Roslyn is invoked as a part of .NET Native toolchain.  This failure does happen often though.

**Comment by jaredpar** [comment]
@AntonLapounov does the optimization you mention ever occur in the CLR JIT? Or is this re-ordering only something that will be observed with .NET Native.

I understand it's a legal optimization. Not questioning that. What I want to understand though is why we've never seen this before. Other race conditions in this layer of the compiler have shown up in testing.

**Comment by AntonLapounov** [comment]
@jaredpar I don't know the answer for CLR JIT.  This failure does *not* happen frequently (sorry, made a typo at first).  We started collecting crash dumps not a long time ago and we had a significantly more frequent race condition in the compiler, which I guess masked this less frequent one.

Please note that comparing `typesNs` local var to `null` is not guaranteed to work correctly, because the compiler is free to re-read `_typesByNS`'s value from memory.  See this [issue](https://github.com/dotnet/corert/pull/4203/commits/e2c7f357fab85bc1acd6661bd511b2428980e686) for an example.

**Comment by jaredpar** [comment]
@AntonLapounov 

> Please note that comparing typesNs local var to null is not guaranteed to work correctly, because the compiler is free to re-read _typesByNS's value from memory.

Correct. But I believe we can still use `_typeByNS` directly in the `null`. That's one of the exceptions in the memory model IIRC. I'm not 100% sure about that though, takes me a while to page that all back into memory.

**Comment by AntonLapounov** [comment]
@jaredpar Like below?  I think it would still be the same issue.  Sorry, not sure what exception you mean.
```C#
var typesByNS = _typesByNS;
if (_typesByNS != null) { LoadAllMembers(typesByNS); … }
```

**Comment by jaredpar** [comment]
@AntonLapounov no. I was thinking of a bit different. The exception I'm referring to is the language to allow for double check locking patterns. It's too prominent in .NET and our memory models have exceptions to make it work. Thinking further I don't believe it applies here.
